cargo-pgrx/src/command/upgrade.rs

```diff
@@ -0,0 +1,216 @@
+                return Err(eyre!(
+                    "Could not parse dependency \
+                entry for {dep_name} due to error: {e}"
+                ))
+            }
+        };
+        let reg_url = registry_url(path, parsed_dep.registry())
+            .map_err(|e| eyre!("Unable to fetch registry URL for path: {e}"))?;
+        let target_version = match self.target_version {
+            Some(ref ver) => Some(ver.clone()),
+            None => cargo_edit::get_latest_dependency(
+                dep_name,
+                self.allow_prerelease,
+                None,
+                path,
+                Some(&reg_url),
+            )
+            .map_err(|e| {
+                eyre!(
+                    "Unable to fetch the latest version \
+                        for crate {dep_name} due to {e}"
+                )
+            })?
+            .version()
+            .map(|s| s.to_string()),
+        };
+        let target_version = match target_version {
+            Some(ver) => ver,
+            None => {
+                return Err(eyre!(
+                    "Unable to update {dep_name} \
+                , no provided crate version and a latest version \
+                could not be retrieved from the registry."
+                ))
+            }
+        };
+
+        if let Some(source) = parsed_dep.source().cloned() {
+            debug!(
+                "Found dependency {dep_name} with current \
+                source {source:#?}"
+            );
+            if let cargo_edit::Source::Workspace(_) = &source {
+                error!(
+                    "Cannot upgrade the version of \
+                        {dep_name} because it is set in the \
+                        workspace, please run `cargo pgrx \
+                        upgrade` in the workspace directory."
+                );
+            } else if source.get_version().is_some() {
+                replace_version(
+                    target_version.as_str(),
+                    path.as_path(),
+                    &mut key,
+                    dep,
+                    parsed_dep,
+                    source,
+                )?;
+            } else {
+                info!("No version specified for {dep_name}, not upgrading.");
+            }
+        }
+        Ok(())
+    }
+}
+
+impl CommandExecute for Upgrade {
+    #[tracing::instrument(level = "error", skip(self))]
+    fn execute(self) -> eyre::Result<()> {
+        const RELEVANT_PACKAGES: [&str; 3] = ["pgrx", "pgrx-macros", "pgrx-tests"];
+        // Canonicalize because cargo-edit does not accept relative paths.
+        let path =
+            std::fs::canonicalize(self.path.clone().unwrap_or(PathBuf::from("./Cargo.toml")))?;
+
+        let mut manifest = cargo_edit::LocalManifest::find(Some(&path))
+            .map_err(|e| eyre!("Error opening manifest: {e}"))?;
+
+        for dep_table in manifest.get_dependency_tables_mut() {
+            for dep_name in RELEVANT_PACKAGES {
```

Comment:
hmm. so, i tried this on a workspace that has pgrx in a *member*, but not in the workspace *root*. right now the way `pgrx upgrade` treats that is to silently do nothing. maybe we could add a warning about that? or even add a `-p` flag, analogous to `cargo build -p`, that says "update the dependencies for this member of the workspace"?

Reply:
I was considering adding an `-r` as in recursive flag along those lines. It might be a good idea. That's a larger refactor of this PR but I'll look into what would be needed.

Reply:
I'd like this to land with the ability to point it at a specific package in a workspace, at least. It doesn't have to have a full-on recursive WalkDir logic.

Reply:
@workingjubilee Should this be `--package` or would that be ambiguous / seem like it's referring to the package to update?

Reply:
Right now I've got an implementation of this, which needs review.

Reply:
> @workingjubilee Should this be --package or would that be ambiguous / seem like it's referring to the package to update?

...oof, uh, idk.

Reply:
I think `--package` is unfortunately probably what people will expect here, even though it's confusing, and we can revise on user experience reports.